text/003-expressing-npm-deps.md

```diff
@@ -0,0 +1,117 @@
+# RFC: npm Dependency Expression
+
+## Summary
+
+Allow for the expression of npm dependencies in Rust-generated WebAssembly projects that use the `wasm-pack` workflow.
+
+## Motivation
+
+In keeping with the team’s goal to allow the surgical replacement of JS codepaths with Rust-generated WebAssembly, developers using the `wasm-pack` workflow should be able to express dependency on packages from the npm registry, other registry, or version-control repository. 
+
+## Guiding Values
+- Development on Rust-generated WebAssembly projects should allow developers to use the development environment they are most comfortable with. Developers writing Rust should get to use Rust, and developers using JavaScript should get to use a JS based runtime environment (Node.js, Chakra, etc).
+- JavaScript tooling and workflows should be usable with Rust-generated WebAssembly projects. For example, bundlers like WebPack and Parcel, or dependency management tools such as `npm audit` and GreenKeeper.
+- When possible, decisions should be made that allow the solution to be available to developers of not just Rust, but also C, and C++.
+- Decisions should be focused on creating workflows that allow developers an easy learning curve and productive development experience.
+
+
+
+## Solutions
+
+Any solution to a problem like this involves 2 steps: 
+  
+1. How to index the third-party dependencies (in this case: npm packages), and 
+2. How to "require" or "import" the packages into code. 
+
+The second of these is simpler than the first so let's start with that: 
+
+### Requiring an npm package
+
+To require an npm package in your Rust code, you will use the `wasm-bindgen`
+attribute, passing in `module = "name-of-package"`.
+
+```rs
+// src/foo.rs
+    
+#[wasm_bindgen(module = "moment")]
+extern {
+  // imported from moment.js
+}
+```
+
+## Indexing the npm packages
+
+This question of how to index, or even if, to index, the npm packages is a large one with
+several considerations. The below options were all considered. We believe that the
+`package.json` solution is the best, at the moment.
+
+### `package.json`
+
+*This is likely the best choice. Although it requires that Rust developers use a* `*package.json*` *file, it allows the best interoperability with existing JavaScript tooling and is agnostic to target language (Rust, C, C++).*
+
+Create a file called `package.json` in the root of your Rust library. Fill out dependencies as per specification: https://docs.npmjs.com/files/package.json#dependencies. You can use `npm install` to add dependencies: Although npm will warn that your `package.json` is missing metadata, it will add the dependency entry.
+
+Note: All meta-data in this file will be ignored during `wasm-pack build` in favor of metadata in the `Cargo.toml`. The potential duplication of metadata is a downside to this solution.
```

Comment:
I think that depends on whether `wasm-pack build` will run a bundler or not (such as Webpack).

Reply:
i'm not sure if it's related to calling a bundler, which i don't think wasm-pack will ever do. can you say more on that @Pauan ?

npm dependencies are not required to be installed in order for the package to be published, though they will be necessary to run tests. i would be inclined to have test run it since it's required, but am not convinced it's necessary for build (it's a long step and should probably only be run if necessary). am open to hearing more examples of the workflows people anticipate/already use, though.

Reply:
@ashleygwilliams My thinking is basically this: if wasm-pack *doesn't* run a bundler (such as Webpack), then the code will not work as-is, so it will require the user to run a bundle step themself.

And since wasm-pack isn't running a bundler, having `wasm-pack build` run `npm install` is unusual, because `npm install` doesn't do anything useful (since the dependencies aren't needed at this point).

So as a user it makes me wonder, "why is wasm-pack running `npm install`? and why isn't it running a bundler?"

I'm neutral on whether `wasm-pack test` should run `npm install` or not, though it's a little weird that it's inconsistent with `wasm-build`

Reply:
I think running `npm install` for `wasm-pack test` but not for `wasm-pack build` makes sense. If we find that users are confused and this is somehow breaking their workflow, then we can revisit the decision, I think.